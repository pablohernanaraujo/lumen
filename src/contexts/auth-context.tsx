import React, {
  createContext,
  type FC,
  type ReactNode,
  useContext,
  useEffect,
  useState,
} from 'react';
import { type User } from '@react-native-google-signin/google-signin';

import { AuthService } from '../services/auth-service';

interface AuthState {
  user: User | null;
  isLoading: boolean;
  isAuthenticated: boolean;
}

export interface AuthContextType extends AuthState {
  signInWithGoogle: () => Promise<void>;
  signOut: () => Promise<void>;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

interface AuthProviderProps {
  children: ReactNode;
}

export const AuthProvider: FC<AuthProviderProps> = ({ children }) => {
  const [authState, setAuthState] = useState<AuthState>({
    user: null,
    isLoading: true,
    isAuthenticated: false,
  });

  // Configure Google Sign-In
  useEffect(() => {
    AuthService.configure();
  }, []);

  // Check for existing authentication on app start
  useEffect(() => {
    const setUnauthenticatedState = (): void => {
      setAuthState({
        user: null,
        isLoading: false,
        isAuthenticated: false,
      });
    };

    const handleGoogleSignInUser = async (currentUser: User): Promise<void> => {
      const tokens = await AuthService.refreshTokens();
      if (tokens) {
        await AuthService.storeAuthData(tokens, currentUser);
      }

      setAuthState({
        user: currentUser,
        isLoading: false,
        isAuthenticated: true,
      });
    };

    const checkStoredData = async (): Promise<void> => {
      const { token, user } = await AuthService.getStoredAuthData();
      if (token && user) {
        console.warn(
          'AsyncStorage has user data but Google Sign-In is not active. Clearing stored data.',
        );
        await AuthService.clearStoredAuthData();
      }
    };

    const checkAuthState = async (): Promise<void> => {
      try {
        // Try to restore session silently from iOS Keychain
        const currentUser = await AuthService.restoreSignInSilently();

        if (currentUser) {
          // Successfully restored from Keychain
          await handleGoogleSignInUser(currentUser);
          return;
        }

        // No silent sign-in possible, check stored data for cleanup
        await checkStoredData();
        setUnauthenticatedState();
      } catch (error) {
        console.error('Error checking auth state:', error);
        await AuthService.clearStoredAuthData();
        setUnauthenticatedState();
      }
    };

    checkAuthState();
  }, []);

  const signInWithGoogle = async (): Promise<void> => {
    try {
      setAuthState((prev) => ({
        ...prev,
        isLoading: true,
      }));

      const { user } = await AuthService.signInWithGoogle();

      setAuthState({
        user,
        isLoading: false,
        isAuthenticated: true,
      });
    } catch (error: unknown) {
      console.error('Google Sign-In error:', error);

      setAuthState((prev) => ({
        ...prev,
        isLoading: false,
      }));

      // Re-throw the error for UI handling
      const error_ =
        error instanceof Error ? error : new Error('Google Sign-In failed');
      throw error_;
    }
  };

  const signOut = async (): Promise<void> => {
    try {
      setAuthState((prev) => ({
        ...prev,
        isLoading: true,
      }));

      await AuthService.signOut();

      setAuthState({
        user: null,
        isLoading: false,
        isAuthenticated: false,
      });
    } catch (error) {
      console.error('Sign out error:', error);

      // Even if sign-out fails, clear local state
      setAuthState({
        user: null,
        isLoading: false,
        isAuthenticated: false,
      });
    }
  };

  const contextValue: AuthContextType = {
    ...authState,
    signInWithGoogle,
    signOut,
  };

  return (
    <AuthContext.Provider value={contextValue}>{children}</AuthContext.Provider>
  );
};

export const useAuth = (): AuthContextType => {
  const context = useContext(AuthContext);

  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }

  return context;
};
