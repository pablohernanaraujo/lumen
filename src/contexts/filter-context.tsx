/* eslint-disable max-statements */
/* eslint-disable complexity */
import React, {
  createContext,
  type FC,
  type ReactNode,
  useContext,
  useEffect,
  useState,
} from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';

import type { CryptoFilters } from '../screens/modals/filter-modal/types';

const FILTERS_STORAGE_KEY = '@lumen_crypto_filters';

const DEFAULT_FILTERS: CryptoFilters = {
  price: {
    enabled: false,
    min: undefined,
    max: undefined,
  },
  marketCap: {
    enabled: false,
    category: undefined,
    min: undefined,
    max: undefined,
  },
  volume: {
    enabled: false,
    min: undefined,
    max: undefined,
  },
  change24h: {
    enabled: false,
    type: undefined,
    min: undefined,
    max: undefined,
  },
  ranking: {
    enabled: false,
    topN: undefined,
  },
  quickFilters: {
    trending: false,
    recentlyAdded: false,
    highVolume: false,
  },
};

interface FilterContextType {
  filters: CryptoFilters;
  isLoading: boolean;
  hasActiveFilters: boolean;
  clearFilters: () => Promise<void>;
  updatePriceFilter: (min?: number, max?: number) => void;
  updateMarketCapFilter: (
    category?: 'small' | 'mid' | 'large' | 'custom',
    min?: number,
    max?: number,
  ) => void;
  updateVolumeFilter: (min?: number, max?: number) => void;
  updateChange24hFilter: (
    type?: 'gainers' | 'losers' | 'custom',
    min?: number,
    max?: number,
  ) => void;
  updateRankingFilter: (topN?: 10 | 50 | 100 | 500) => void;
  toggleQuickFilter: (
    filterName: 'trending' | 'recentlyAdded' | 'highVolume',
  ) => void;
  getActiveFilterCount: () => number;
  applyFiltersToData: <
    T extends {
      current_price: number;
      market_cap: number;
      total_volume: number;
      price_change_percentage_24h: number;
      market_cap_rank: number;
    },
  >(
    data: T[],
  ) => T[];
  saveFilters: (newFilters: CryptoFilters) => Promise<void>;
}

const FilterContext = createContext<FilterContextType | undefined>(undefined);

export const FilterProvider: FC<{ children: ReactNode }> = ({ children }) => {
  const [filters, setFilters] = useState<CryptoFilters>(DEFAULT_FILTERS);
  const [isLoading, setIsLoading] = useState(true);
  const [hasActiveFilters, setHasActiveFilters] = useState(false);

  // Load saved filters on mount
  useEffect(() => {
    loadFilters();
  }, []);

  // Update hasActiveFilters when filters change
  useEffect(() => {
    const active = checkActiveFilters(filters);
    setHasActiveFilters(active);
  }, [filters]);

  const loadFilters = async (): Promise<void> => {
    try {
      const savedFilters = await AsyncStorage.getItem(FILTERS_STORAGE_KEY);
      if (savedFilters) {
        setFilters(JSON.parse(savedFilters));
      }
    } catch (error) {
      console.error('Error loading filters:', error);
    } finally {
      setIsLoading(false);
    }
  };

  const saveFilters = async (newFilters: CryptoFilters): Promise<void> => {
    try {
      await AsyncStorage.setItem(
        FILTERS_STORAGE_KEY,
        JSON.stringify(newFilters),
      );
      setFilters(newFilters);
    } catch (error) {
      console.error('Error saving filters:', error);
    }
  };

  const checkActiveFilters = (currentFilters: CryptoFilters): boolean => {
    if (currentFilters.price?.enabled) return true;
    if (currentFilters.marketCap?.enabled) return true;
    if (currentFilters.volume?.enabled) return true;
    if (currentFilters.change24h?.enabled) return true;
    if (currentFilters.ranking?.enabled) return true;
    if (currentFilters.quickFilters?.trending) return true;
    if (currentFilters.quickFilters?.recentlyAdded) return true;
    if (currentFilters.quickFilters?.highVolume) return true;
    return false;
  };

  const clearFilters = async (): Promise<void> => {
    await saveFilters(DEFAULT_FILTERS);
  };

  const updatePriceFilter = (min?: number, max?: number): void => {
    const newFilters = {
      ...filters,
      price: {
        enabled: min !== undefined || max !== undefined,
        min,
        max,
      },
    };
    saveFilters(newFilters);
  };

  const updateMarketCapFilter = (
    category?: 'small' | 'mid' | 'large' | 'custom',
    min?: number,
    max?: number,
  ): void => {
    const newFilters = {
      ...filters,
      marketCap: {
        enabled: true,
        category,
        min,
        max,
      },
    };
    saveFilters(newFilters);
  };

  const updateVolumeFilter = (min?: number, max?: number): void => {
    const newFilters = {
      ...filters,
      volume: {
        enabled: min !== undefined || max !== undefined,
        min,
        max,
      },
    };
    saveFilters(newFilters);
  };

  const updateChange24hFilter = (
    type?: 'gainers' | 'losers' | 'custom',
    min?: number,
    max?: number,
  ): void => {
    const newFilters = {
      ...filters,
      change24h: {
        enabled: true,
        type,
        min,
        max,
      },
    };
    saveFilters(newFilters);
  };

  const updateRankingFilter = (topN?: 10 | 50 | 100 | 500): void => {
    const newFilters = {
      ...filters,
      ranking: {
        enabled: topN !== undefined,
        topN,
      },
    };
    saveFilters(newFilters);
  };

  const toggleQuickFilter = (
    filterName: 'trending' | 'recentlyAdded' | 'highVolume',
  ): void => {
    const newFilters = {
      ...filters,
      quickFilters: {
        ...filters.quickFilters,
        [filterName]: !filters.quickFilters?.[filterName],
      },
    };
    saveFilters(newFilters);
  };

  const getActiveFilterCount = (): number => {
    let count = 0;
    if (filters.price?.enabled) count++;
    if (filters.marketCap?.enabled) count++;
    if (filters.volume?.enabled) count++;
    if (filters.change24h?.enabled) count++;
    if (filters.ranking?.enabled) count++;
    if (filters.quickFilters?.trending) count++;
    if (filters.quickFilters?.recentlyAdded) count++;
    if (filters.quickFilters?.highVolume) count++;
    return count;
  };

  const applyFiltersToData = <
    T extends {
      current_price: number;
      market_cap: number;
      total_volume: number;
      price_change_percentage_24h: number;
      market_cap_rank: number;
    },
  >(
    data: T[],
  ): T[] => {
    let filteredData = [...data];

    // Apply price filter
    if (filters.price?.enabled) {
      filteredData = filteredData.filter((item) => {
        if (
          filters.price?.min !== undefined &&
          item.current_price < filters.price.min
        ) {
          return false;
        }
        if (
          filters.price?.max !== undefined &&
          item.current_price > filters.price.max
        ) {
          return false;
        }
        return true;
      });
    }

    // Apply market cap filter
    if (filters.marketCap?.enabled) {
      filteredData = filteredData.filter((item) => {
        if (filters.marketCap?.category) {
          switch (filters.marketCap.category) {
            case 'small':
              return item.market_cap < 1000000000; // < $1B
            case 'mid':
              return (
                item.market_cap >= 1000000000 && item.market_cap < 10000000000
              ); // $1B - $10B
            case 'large':
              return item.market_cap >= 10000000000; // >= $10B
            case 'custom':
              if (
                filters.marketCap?.min !== undefined &&
                item.market_cap < filters.marketCap.min
              ) {
                return false;
              }
              if (
                filters.marketCap?.max !== undefined &&
                item.market_cap > filters.marketCap.max
              ) {
                return false;
              }
              return true;
            default:
              return true;
          }
        }
        return true;
      });
    }

    // Apply volume filter
    if (filters.volume?.enabled) {
      filteredData = filteredData.filter((item) => {
        if (
          filters.volume?.min !== undefined &&
          item.total_volume < filters.volume.min
        ) {
          return false;
        }
        if (
          filters.volume?.max !== undefined &&
          item.total_volume > filters.volume.max
        ) {
          return false;
        }
        return true;
      });
    }

    // Apply 24h change filter
    if (filters.change24h?.enabled) {
      filteredData = filteredData.filter((item) => {
        if (filters.change24h?.type) {
          switch (filters.change24h.type) {
            case 'gainers':
              return item.price_change_percentage_24h > 0;
            case 'losers':
              return item.price_change_percentage_24h < 0;
            case 'custom':
              if (
                filters.change24h?.min !== undefined &&
                item.price_change_percentage_24h < filters.change24h.min
              ) {
                return false;
              }
              if (
                filters.change24h?.max !== undefined &&
                item.price_change_percentage_24h > filters.change24h.max
              ) {
                return false;
              }
              return true;
            default:
              return true;
          }
        }
        return true;
      });
    }

    // Apply ranking filter
    if (filters.ranking?.enabled && filters.ranking?.topN) {
      filteredData = filteredData.filter(
        (item) => item.market_cap_rank <= (filters.ranking?.topN || 100),
      );
    }

    // Apply quick filters
    if (filters.quickFilters?.highVolume) {
      // Filter for high volume (top 20% by volume)
      const sortedByVolume = [...filteredData].sort(
        (a, b) => b.total_volume - a.total_volume,
      );
      const top20Percent = Math.ceil(sortedByVolume.length * 0.2);
      const highVolumeThreshold =
        sortedByVolume[top20Percent - 1]?.total_volume || 0;
      filteredData = filteredData.filter(
        (item) => item.total_volume >= highVolumeThreshold,
      );
    }

    return filteredData;
  };

  const value: FilterContextType = {
    filters,
    isLoading,
    hasActiveFilters,
    clearFilters,
    updatePriceFilter,
    updateMarketCapFilter,
    updateVolumeFilter,
    updateChange24hFilter,
    updateRankingFilter,
    toggleQuickFilter,
    getActiveFilterCount,
    applyFiltersToData,
    saveFilters,
  };

  return (
    <FilterContext.Provider value={value}>{children}</FilterContext.Provider>
  );
};

export const useFilters = (): FilterContextType => {
  const context = useContext(FilterContext);
  if (context === undefined) {
    throw new Error('useFilters must be used within a FilterProvider');
  }
  return context;
};
