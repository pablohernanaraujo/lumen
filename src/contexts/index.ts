export type { AuthContextType } from './auth-context';
export { AuthProvider, useAuth } from './auth-context';
