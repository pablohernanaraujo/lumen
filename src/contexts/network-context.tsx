import {
  createContext,
  type FC,
  type ReactNode,
  useContext,
  useEffect,
  useState,
} from 'react';

import {
  type NetworkInfo,
  networkStrategyService,
} from '../services/network-strategy-service';

export interface NetworkContextValue {
  isConnected: boolean;
  isReachable: boolean;
  networkType: string;
  isWiFi: boolean;
  isCellular: boolean;
  isExpensive: boolean;
  connectionQuality?: string;
  isOnline: boolean;
  wasOffline: boolean;
}

const NetworkContext = createContext<NetworkContextValue | undefined>(
  undefined,
);

export const useNetworkStatus = (): NetworkContextValue => {
  const context = useContext(NetworkContext);
  if (context === undefined) {
    throw new Error('useNetworkStatus must be used within a NetworkProvider');
  }
  return context;
};

interface NetworkProviderProps {
  children: ReactNode;
}

export const NetworkProvider: FC<NetworkProviderProps> = ({ children }) => {
  const [networkInfo, setNetworkInfo] = useState<NetworkInfo>(
    networkStrategyService.getCurrentNetworkInfo(),
  );
  const [wasOffline, setWasOffline] = useState(false);

  useEffect(() => {
    const unsubscribe = networkStrategyService.onNetworkChange(
      (info: NetworkInfo) => {
        const previouslyConnected = networkInfo.isConnected;
        const nowConnected = info.isConnected && info.isInternetReachable;

        if (!previouslyConnected && nowConnected) {
          setWasOffline(true);
          setTimeout(() => setWasOffline(false), 3000);
        }

        setNetworkInfo(info);
      },
    );

    return unsubscribe;
  }, [networkInfo.isConnected]);

  const isOnline = networkInfo.isConnected && networkInfo.isInternetReachable;

  const value: NetworkContextValue = {
    isConnected: networkInfo.isConnected,
    isReachable: networkInfo.isInternetReachable,
    networkType: networkInfo.type,
    isWiFi: networkInfo.isWiFi,
    isCellular: networkInfo.isCellular,
    isExpensive: networkInfo.isExpensive,
    connectionQuality: networkInfo.effectiveConnectionType,
    isOnline,
    wasOffline,
  };

  return (
    <NetworkContext.Provider value={value}>{children}</NetworkContext.Provider>
  );
};
