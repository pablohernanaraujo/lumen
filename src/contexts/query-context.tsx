/* eslint-disable max-statements */
/* eslint-disable complexity */
import React, { type FC, type ReactNode, useEffect } from 'react';
import { AppState, type AppStateStatus } from 'react-native';
import NetInfo from '@react-native-community/netinfo';
import {
  focusManager,
  onlineManager,
  QueryClient,
  QueryClientProvider,
} from '@tanstack/react-query';

interface QueryProviderProps {
  children: ReactNode;
}

// Network-aware retry function
const networkAwareRetry = (failureCount: number, error: unknown): boolean => {
  // Don't retry on client errors (4xx) except 429 (rate limit)
  if (error && typeof error === 'object' && 'response' in error) {
    const response = (error as { response?: { status?: number } }).response;
    if (response?.status) {
      const status = response.status;
      // Don't retry on 4xx errors except 429 (Too Many Requests)
      if (status >= 400 && status < 500 && status !== 429) {
        return false;
      }
      // Always retry on 5xx server errors
      if (status >= 500) {
        return failureCount < 3; // More retries for server errors
      }
    }
  }

  // Retry on network errors, timeouts, etc.
  if (error && typeof error === 'object') {
    const errorName = (error as { name?: string }).name;
    const errorCode = (error as { code?: string }).code;

    if (
      errorName === 'NetworkError' ||
      errorName === 'TimeoutError' ||
      errorCode === 'ECONNABORTED' ||
      errorCode === 'ENOTFOUND'
    ) {
      return failureCount < 3;
    }
  }

  return failureCount < 2; // Default retry count
};

// Smart stale time based on data type
const getStaleTime = (queryKey: readonly unknown[]): number => {
  const key = String(queryKey[0] || '');

  // Market data - 2 minutes (frequently changing prices)
  if (key.includes('crypto-list') || key.includes('price')) {
    return 2 * 60 * 1000;
  }

  // Detailed coin data - 5 minutes (less frequently changing)
  if (key.includes('crypto-detail')) {
    return 5 * 60 * 1000;
  }

  // Search results - 10 minutes (rarely changing)
  if (key.includes('crypto-search')) {
    return 10 * 60 * 1000;
  }

  // User-specific data - 1 minute (frequently changing)
  if (key.includes('user') || key.includes('portfolio')) {
    return 1 * 60 * 1000;
  }

  // Default - 5 minutes
  return 5 * 60 * 1000;
};

// Smart cache time based on data type and network conditions
const getCacheTime = (queryKey: readonly unknown[]): number => {
  const key = String(queryKey[0] || '');

  // Critical data - keep longer in cache
  if (key.includes('crypto-list') || key.includes('crypto-detail')) {
    return 30 * 60 * 1000; // 30 minutes
  }

  // Search results - medium cache time
  if (key.includes('crypto-search')) {
    return 20 * 60 * 1000; // 20 minutes
  }

  // User data - shorter cache time
  if (key.includes('user') || key.includes('portfolio')) {
    return 5 * 60 * 1000; // 5 minutes
  }

  // Default - 15 minutes
  return 15 * 60 * 1000;
};

// Create a client with optimized settings for React Native
const createQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        // Default stale time (will be overridden in individual hooks)
        staleTime: 5 * 60 * 1000, // 5 minutes default
        // Default cache time (will be overridden in individual hooks)
        gcTime: 15 * 60 * 1000, // 15 minutes default
        // Network-aware retry logic
        retry: networkAwareRetry,
        // Progressive retry delay with jitter
        retryDelay: (attemptIndex, error) => {
          const baseDelay = Math.min(1000 * Math.pow(2, attemptIndex), 30000);
          const jitter = Math.random() * 1000;

          // Longer delays for rate limit errors
          if (error && typeof error === 'object' && 'response' in error) {
            const response = (error as { response?: { status?: number } })
              .response;
            if (response?.status === 429) {
              return baseDelay * 2 + jitter; // Double delay for rate limits
            }
          }

          return baseDelay + jitter;
        },
        // Smart refetch behavior
        refetchOnWindowFocus: false, // Disabled for mobile app
        refetchOnReconnect: 'always', // Always refetch when network reconnects
        refetchOnMount: (query: {
          queryKey: readonly unknown[];
          state: { dataUpdatedAt: number };
        }) => {
          // Only refetch on mount if data is stale
          const now = Date.now();
          const staleTime = getStaleTime(query.queryKey);
          const dataUpdatedAt = query.state.dataUpdatedAt;

          return now - dataUpdatedAt > staleTime;
        },
        // Background refetch settings
        refetchInterval: false, // Disabled by default
        refetchIntervalInBackground: false,
        // Network mode for offline support
        networkMode: 'offlineFirst',
      },
      mutations: {
        // More aggressive retry for mutations
        retry: (failureCount, error) => {
          // Don't retry client errors except rate limits
          if (error && typeof error === 'object' && 'response' in error) {
            const response = (error as { response?: { status?: number } })
              .response;
            if (
              response?.status &&
              response.status >= 400 &&
              response.status < 500 &&
              response.status !== 429
            ) {
              return false;
            }
          }
          return failureCount < 2;
        },
        // Progressive retry delay for mutations
        retryDelay: (attemptIndex) =>
          Math.min(1000 * Math.pow(2, attemptIndex), 10000),
        // Network mode for mutations
        networkMode: 'online',
      },
    },
  });

// Create a single instance to avoid creating multiple clients
let queryClientInstance: QueryClient | null = null;

const getQueryClient = (): QueryClient => {
  if (!queryClientInstance) {
    queryClientInstance = createQueryClient();
  }
  return queryClientInstance;
};

export const QueryProvider: FC<QueryProviderProps> = ({ children }) => {
  const queryClient = getQueryClient();

  useEffect(() => {
    // Set up network state management
    const setupNetworkManagement = async (): Promise<() => void> => {
      try {
        // Subscribe to network state changes
        const unsubscribe = NetInfo.addEventListener((state) => {
          onlineManager.setOnline(
            state.isConnected != null &&
              state.isConnected &&
              Boolean(state.isInternetReachable),
          );
        });

        // Set initial network state
        const netInfoState = await NetInfo.fetch();
        onlineManager.setOnline(
          netInfoState.isConnected != null &&
            netInfoState.isConnected &&
            Boolean(netInfoState.isInternetReachable),
        );

        return unsubscribe;
      } catch {
        console.warn(
          '[QueryProvider] NetInfo not available, using default online state',
        );
        // Fallback to assuming online
        onlineManager.setOnline(true);
        return () => {};
      }
    };

    setupNetworkManagement().catch((error) => {
      console.warn(
        '[QueryProvider] Failed to setup network management:',
        error,
      );
    });
  }, []);

  useEffect(() => {
    // Set up focus management for app state changes
    const handleAppStateChange = (status: AppStateStatus): void => {
      // Disable focus manager when app goes to background
      // Enable when app becomes active
      focusManager.setFocused(status === 'active');

      // Optionally pause/resume queries based on app state
      if (status === 'background') {
        // Optionally cancel ongoing queries when app goes to background
        queryClient.cancelQueries();
      }
    };

    const subscription = AppState.addEventListener(
      'change',
      handleAppStateChange,
    );

    // Set initial focus state
    focusManager.setFocused(AppState.currentState === 'active');

    return () => subscription?.remove();
  }, [queryClient]);

  useEffect(() => {
    // Set up periodic cache cleanup
    const cleanupInterval = setInterval(
      () => {
        // Remove expired queries
        queryClient.getQueryCache().clear();

        // Optionally run garbage collection
        queryClient.invalidateQueries({
          predicate: (query: {
            queryKey: readonly unknown[];
            state: { dataUpdatedAt: number };
          }) => {
            const now = Date.now();
            const cacheTime = getCacheTime(query.queryKey);
            return now - query.state.dataUpdatedAt > cacheTime * 2; // Double the cache time for cleanup
          },
        });
      },
      15 * 60 * 1000,
    ); // Clean up every 15 minutes

    return () => clearInterval(cleanupInterval);
  }, [queryClient]);

  useEffect(() => {
    // Set up intelligent background refetch for critical data
    const backgroundRefetchInterval = setInterval(
      async () => {
        // Only refetch if app is in foreground and online
        if (AppState.currentState === 'active' && onlineManager.isOnline()) {
          try {
            const netInfo = await NetInfo.fetch();

            // More aggressive refetch on WiFi, conservative on cellular
            const isWiFi = netInfo.type === 'wifi';
            const refetchKey = isWiFi ? 'crypto-list' : null;

            if (refetchKey) {
              queryClient.invalidateQueries({
                predicate: (query: { queryKey: readonly unknown[] }) =>
                  String(query.queryKey[0] || '').includes(refetchKey),
              });
            }
          } catch {
            // Fallback behavior if NetInfo is not available
            console.warn(
              '[QueryProvider] NetInfo not available for background refetch',
            );
          }
        }
      },
      15 * 60 * 1000,
    ); // Check every 15 minutes (reduced from 5 minutes to prevent rate limits)

    return () => clearInterval(backgroundRefetchInterval);
  }, [queryClient]);

  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

// Export utilities for external use
export { getCacheTime, getStaleTime, networkAwareRetry };
