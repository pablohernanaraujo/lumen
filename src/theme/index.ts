export { colorTokens } from './colors';
export { makeStyles } from './make-styles';
export { ThemeProvider, useTheme } from './theme-provider';
export { baseTokens } from './tokens';
export type {
  BaseTokens,
  ColorTokens,
  EmphasisLevel,
  Mode,
  NamedStyles,
  StyleFunction,
  StylesOrFunction,
  Theme,
  ThemeColors,
  ThemeContextValue,
  ThemeProviderProps,
} from './types';
export {
  createButtonStyle,
  createCardStyle,
  createInputStyle,
  createShadow,
  createTextStyle,
  getColorOpacity,
  getContrastColor,
  isLightColor,
} from './utils';
