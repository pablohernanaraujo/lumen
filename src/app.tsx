import { type FC, StrictMode } from 'react';
import { StatusBar, View } from 'react-native';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { NavigationContainer } from '@react-navigation/native';

import {
  AuthProvider,
  FilterProvider,
  NetworkProvider,
  QueryProvider,
} from './contexts';
import { useAppInitialization } from './hooks/use-app-initialization';
import { RootNavigator } from './routing';
import { navigationRef } from './routing/routing-service';
import { ThemeProvider, useTheme } from './theme';
import { ErrorState, LoadingIndicator, NetworkBanner } from './ui';

const AppContent: FC = () => {
  const { theme, mode } = useTheme();
  const { isInitialized, isInitializing, error } = useAppInitialization();

  if (isInitializing) {
    return (
      <GestureHandlerRootView style={{ flex: 1 }}>
        <SafeAreaProvider>
          <StatusBar
            barStyle={mode === 'dark' ? 'light-content' : 'dark-content'}
            backgroundColor={theme.colors.background}
          />
          <View
            style={{
              flex: 1,
              justifyContent: 'center',
              alignItems: 'center',
              backgroundColor: theme.colors.background,
            }}
          >
            <LoadingIndicator label="Initializing app..." showLabel />
          </View>
        </SafeAreaProvider>
      </GestureHandlerRootView>
    );
  }

  if (error) {
    return (
      <GestureHandlerRootView style={{ flex: 1 }}>
        <SafeAreaProvider>
          <StatusBar
            barStyle={mode === 'dark' ? 'light-content' : 'dark-content'}
            backgroundColor={theme.colors.background}
          />
          <View
            style={{
              flex: 1,
              backgroundColor: theme.colors.background,
            }}
          >
            <ErrorState
              title="Initialization Failed"
              message={error}
              retryText="Retry"
              onRetry={(): void => {
                // Force app reload by reloading the bundle
                console.log(
                  'User requested app restart after initialization failure',
                );
              }}
            />
          </View>
        </SafeAreaProvider>
      </GestureHandlerRootView>
    );
  }

  if (!isInitialized) {
    return (
      <GestureHandlerRootView style={{ flex: 1 }}>
        <SafeAreaProvider>
          <StatusBar
            barStyle={mode === 'dark' ? 'light-content' : 'dark-content'}
            backgroundColor={theme.colors.background}
          />
          <View
            style={{
              flex: 1,
              backgroundColor: theme.colors.background,
            }}
          >
            <ErrorState
              title="App Not Ready"
              message="App initialization incomplete"
              retryText="Retry"
              onRetry={(): void => {
                console.log(
                  'User requested retry after incomplete initialization',
                );
              }}
            />
          </View>
        </SafeAreaProvider>
      </GestureHandlerRootView>
    );
  }

  return (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <SafeAreaProvider>
        <StatusBar
          barStyle={mode === 'dark' ? 'light-content' : 'dark-content'}
          backgroundColor={theme.colors.background}
        />
        <NavigationContainer ref={navigationRef}>
          <RootNavigator />
        </NavigationContainer>
        <NetworkBanner />
      </SafeAreaProvider>
    </GestureHandlerRootView>
  );
};

export const App: FC = () => (
  <StrictMode>
    <QueryProvider>
      <ThemeProvider>
        <NetworkProvider>
          <AuthProvider>
            <FilterProvider>
              <AppContent />
            </FilterProvider>
          </AuthProvider>
        </NetworkProvider>
      </ThemeProvider>
    </QueryProvider>
  </StrictMode>
);
