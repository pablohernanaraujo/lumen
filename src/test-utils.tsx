import React, { type FC, type ReactElement } from 'react';
import {
  render,
  type RenderOptions,
  type RenderResult,
} from '@testing-library/react-native';

import { ThemeProvider } from './theme';
import type { Mode } from './theme/types';

interface AllTheProvidersProps {
  children: React.ReactNode;
  mode?: Mode;
}

const AllTheProviders: FC<AllTheProvidersProps> = ({
  children,
  mode = 'light',
}) => <ThemeProvider preference={mode}>{children}</ThemeProvider>;

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  mode?: Mode;
  wrapper?: React.ComponentType<{ children: React.ReactNode }>;
}

export const customRender = (
  ui: ReactElement,
  { mode = 'light', wrapper, ...renderOptions }: CustomRenderOptions = {},
): RenderResult => {
  const Wrapper: FC<{ children: React.ReactNode }> = ({ children }) => {
    const content = <AllTheProviders mode={mode}>{children}</AllTheProviders>;

    return wrapper
      ? React.createElement(wrapper, { children: content })
      : content;
  };

  return render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });
};

// Re-export everything from React Native Testing Library
export * from '@testing-library/react-native';

// Override the default render with our custom one
export { customRender as render };

// Create snapshot matcher helper
export const createSnapshot = (
  component: ReactElement,
  mode: Mode = 'light',
): void => {
  const { toJSON } = customRender(component, { mode });
  expect(toJSON()).toMatchSnapshot();
};

// Create both light and dark mode snapshots
export const createDualSnapshots = (component: ReactElement): void => {
  describe('snapshots', () => {
    it('should render correctly in light mode', () => {
      createSnapshot(component, 'light');
    });

    it('should render correctly in dark mode', () => {
      createSnapshot(component, 'dark');
    });
  });
};

// Export TestWrapper for use in tests
export const TestWrapper: FC<{ children: React.ReactNode; mode?: Mode }> = ({
  children,
  mode = 'light',
}) => <AllTheProviders mode={mode}>{children}</AllTheProviders>;
