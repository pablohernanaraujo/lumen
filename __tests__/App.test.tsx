import React, { type FC, StrictMode } from 'react';
import { View } from 'react-native';
import { AuthProvider, FilterProvider, QueryProvider } from '../src/contexts';
import { ThemeProvider } from '../src/theme';
import { render, createDualSnapshots } from '../src/test-utils';

// Simple test version of App without complex navigation
const TestApp: FC = () => (
  <StrictMode>
    <QueryProvider>
      <ThemeProvider>
        <AuthProvider>
          <FilterProvider>
            <View testID="app-content">
              <View testID="test-view" />
            </View>
          </FilterProvider>
        </AuthProvider>
      </ThemeProvider>
    </QueryProvider>
  </StrictMode>
);

describe('App', () => {
  createDualSnapshots(<TestApp />);

  describe('rendering', () => {
    it('should render without crashing', () => {
      const { toJSON } = render(<TestApp />);
      expect(toJSON()).toBeTruthy();
    });

    it('should render correctly in light mode', () => {
      const { toJSON } = render(<TestApp />, { mode: 'light' });
      expect(toJSON()).toMatchSnapshot('light-mode');
    });

    it('should render correctly in dark mode', () => {
      const { toJSON } = render(<TestApp />, { mode: 'dark' });
      expect(toJSON()).toMatchSnapshot('dark-mode');
    });
  });

  describe('structure', () => {
    it('should render app structure correctly', () => {
      const { toJSON } = render(<TestApp />);

      // This test verifies the app structure through snapshot
      expect(toJSON()).toBeTruthy();
      expect(toJSON()).toMatchSnapshot('app-structure');
    });
  });
});
