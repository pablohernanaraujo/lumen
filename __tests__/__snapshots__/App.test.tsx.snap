// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App rendering should render correctly in dark mode: dark-mode 1`] = `
<View
  testID="app-content"
>
  <View
    testID="test-view"
  />
</View>
`;

exports[`App rendering should render correctly in light mode: light-mode 1`] = `
<View
  testID="app-content"
>
  <View
    testID="test-view"
  />
</View>
`;

exports[`App snapshots should render correctly in dark mode 1`] = `
<View
  testID="app-content"
>
  <View
    testID="test-view"
  />
</View>
`;

exports[`App snapshots should render correctly in light mode 1`] = `
<View
  testID="app-content"
>
  <View
    testID="test-view"
  />
</View>
`;

exports[`App structure should render app structure correctly: app-structure 1`] = `
<View
  testID="app-content"
>
  <View
    testID="test-view"
  />
</View>
`;
